README.md
# Python Cron Expression
[![PyPI version](https://badge.fury.io/py/PyCronExpression.svg)](https://badge.fury.io/py/PyCronExpression)
![Coverage](https://raw.githubusercontent.com/vubon/PyCronExpression/master/docs/coverage.svg)
[![GitHub version](https://badge.fury.io/gh/vubon%2FPyCronExpression.svg)](https://badge.fury.io/gh/vubon%2FPyCronExpression)

## Introduction
Introducing `PyCronExpression` – a user-friendly Python library for crafting Cron job expressions effortlessly.
With this tool, users can create Cron job expressions by simply passing a timestamp or datetime object.
Say goodbye to the complexities of writing Cron expressions
and embrace a more convenient way to manage scheduled tasks on Linux or AWS platforms.

## Quickstart
### Installation
Install from pypi: 
```shell script
pip install PyCronExpression
```

##### Feature List

| Name                         | status |
|------------------------------|--------|
| Timestamp or datetime object | `done` |
| Minute                       | `done` |
| Hourly                       | `done` |
| Daily                        | `Done` |
| Weekly                       | `done` |

## Example 

#### Timestamp or datetime object
```python
from datetime import datetime
from py_cron_expression import CronJobExpression
cron_job = CronJobExpression()

timestamp = int(datetime.now().timestamp())
res = cron_job.cron_expression(time=timestamp)
print(res)
# Output 
# 10 55 13 2 5 ? 2020

# If you want to remove second from in your cron job expression. 
# Then you need to add cancel='second'
# example
cron_job.cron_expression(time=timestamp, cancel='second')
# Output 
# 55 13 2 5 ? 2020
```
To learn more [Documentation](./docs/guide.md).

## Changelog
See [Changelog](CHANGELOG.md)

## License
MIT

CHANGELOG.md
## Changelog
All notable changes to this project will be documented in this file. <br/>
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.0.1 - 2023-10-28
### Update
- Update docs

## 1.0.0 - 2023-10-28
### Add
- Add weekly expression

## 0.1.0 - 2020-05-28
### Fixed
- Tuple comprehension in validations 

## 0.1.1 - 2020-05-28
### Fixed
- Fixed broken coverage svg 

tests/cron_test.py
import unittest

from py_cron_expression import CronJobExpression


class CronTest(unittest.TestCase):

    def setUp(self) -> None:
        self.valid_timestamp = 1588406110
        self.valid_datetime = ''
        self.invalid_data = "This is a invalid data"
        self.cron_job = CronJobExpression()

    def test_valid_timestamp(self):
        response = self.cron_job.cron_expression(time=self.valid_timestamp)
        self.assertEqual(response, "10 55 13 2 5 ? 2020")

    def test_invalid_data(self):
        self.assertRaises(ValueError, lambda: self.cron_job.cron_expression(time=self.invalid_data))


if __name__ == "__main__":
    unittest.main()


tests/test_daily.py
import unittest

from py_cron_expression import CronJobExpression


class TestDaily(unittest.TestCase):

    def setUp(self) -> None:
        self.cron = CronJobExpression()
        self.valid = {"platform": "aws", "hours": 1}
        self.invalid_platform = {"platform": "awss", "hours": 1}
        self.invalid_hours = {"platform": "aws", "hours": 0}

    def test_valid(self):
        res = self.cron.daily(**self.valid)
        self.assertEqual(res, '0 1 * * ? *')

    def test_invalid_platform(self):
        self.assertRaises(ValueError, lambda: self.cron.daily(**self.invalid_platform))

    def test_invalid_hours(self):
        self.assertRaises(ValueError, lambda: self.cron.daily(**self.invalid_hours))

    def test_without_hours(self):
        self.assertRaises(ValueError, lambda: self.cron.daily(platform='aws'))


tests/test_hourly.py
import unittest

from py_cron_expression import CronJobExpression


class TestHourly(unittest.TestCase):

    def setUp(self) -> None:
        self.cron = CronJobExpression()
        self.valid = {"platform": "aws", "hours": 1}
        self.invalid_platform = {"platform": "awss", "hours": 1}
        self.invalid_hours = {"platform": "aws", "hours": -1}

    def test_valid(self):
        res = self.cron.hourly(**self.valid)
        self.assertEqual(res, '* */1 * * ? *')

    def test_invalid_platform(self):
        self.assertRaises(ValueError, lambda: self.cron.hourly(**self.invalid_platform))

    def test_invalid_hours(self):
        self.assertRaises(ValueError, lambda: self.cron.hourly(**self.invalid_hours))

    def test_without_hours(self):
        self.assertRaises(ValueError, lambda: self.cron.hourly(platform='aws'))


py_cron_expression/util.py
"""
@since: 26 May 2020

# To define the time you can provide concrete values for
# minute (m), hour (h), day of month (dom), month (mon),
# and day of week (dow) or use '*' in these fields (for 'any').#
# Notice that tasks will be started based on the cron's system
# daemon's notion of time and timezones.
Linux platform cron job definition
# .---------------- minute (0 - 59)
# |  .------------- hour (0 - 23)
# |  |  .---------- day of month (1 - 31)
# |  |  |  .------- month (1 - 12) OR jan,feb,mar,apr ...
# |  |  |  |  .---- day of week (0 - 6) (Sunday=0 or 7)
# |  |  |  |  |
# *  *  *  *  *

AWS platform cron job definition
# .---------------------- minute (0 - 59)
# |  .------------------- hour (0 - 23)
# |  |   .---------------- day of month (1 - 31)
# |  |  |   .------------- month (1 - 12) OR jan,feb,mar,apr ...
# |  |  |  |   .---------- day of week (0 - 6) (Sunday=0 or 7)
# |  |  |  |  |  .-------- year
# |  |  |  |  | |
# *  *  *  *  ? *
aws docs link: https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html
"""

PLATFORMS = {
    "linux": "* * * * *",
    "aws": "* * * * ? *"
}

WEEKS = {
    "sunday": "0",
    "monday": "1",
    "tuesday": "2",
    "wednesday": "3",
    "thursday": "4",
    "friday": "5",
    "saturday": "6"
}


def remove_zero(number: str) -> str:
    """
    :param number:
    :return:
    """
    if number.startswith("0") and number.endswith("0"):
        return "0"
    elif number.startswith("0"):
        return number.replace("0", "")
    else:
        return number


py_cron_expression/validations.py
"""
A simple data validation
@Since: 02 May, 2020
"""
from datetime import datetime
from typing import Callable
import pytz
from tzlocal import get_localzone_name

from py_cron_expression.util import WEEKS

GET_TIMEZONE = get_localzone_name()

__all__ = [
    "date_valid", "hour_valid", "minute_valid", "platform_valid", "weekly_valid"
]


def date_valid(func: Callable) -> Callable:
    """
    :param func:
    :return:
    """

    def wrapper(*args, **kwargs):
        """
        :param args:
        :param kwargs:
        :return:
        """
        try:
            time = kwargs['time']
            timezone = kwargs.get('timezone', GET_TIMEZONE)
            cancel = kwargs.get('cancel', None)
            if isinstance(time, int):
                string = datetime.fromtimestamp(time, tz=pytz.timezone(timezone)).strftime('%Y-%m-%dT%H:%M:%S')
                if cancel != 'second':
                    kwargs['cancel'] = False
                kwargs['time'] = string
                return func(*args, **kwargs)
            elif isinstance(time, datetime):
                string = (time.astimezone(tz=pytz.timezone(timezone))).strftime('%Y-%m-%dT%H:%M:%S')
                if cancel != "second":
                    kwargs['cancel'] = False
                kwargs['time'] = string
                return func(*args, **kwargs)
            else:
                raise ValueError(f"Invalid data type {time}")
        except KeyError as key:
            raise KeyError(f"Key name should {key}")

    return wrapper


def hour_valid(func: Callable) -> Callable:
    def wrapper(*args, **kwargs):
        hours = kwargs.get("hours")
        if hours:
            if hours not in tuple(hour for hour in range(1, 24)):
                raise ValueError("Should give valid hours value")
        return func(*args, **kwargs)

    return wrapper


def minute_valid(func: Callable) -> Callable:
    def wrapper(*args, **kwargs):
        minutes = kwargs.get("minutes")
        if minutes and minutes not in tuple(hour for hour in range(1, 60)):
            raise ValueError("Should give valid minutes value")
        return func(*args, **kwargs)

    return wrapper


def weekly_valid(func: Callable) -> Callable:
    def wrapper(*args, **kwargs):
        day = kwargs.get("weekly")
        value = WEEKS.get(day.lower())
        if day and value is None:
            raise ValueError("Should give valid weekly value")
        kwargs["weekly"] = value
        return func(*args, **kwargs)

    return wrapper


def platform_valid(func: Callable) -> Callable:
    def wrapper(*args, **kwargs):
        platform = kwargs.get("platform")
        if not platform:
            raise KeyError("Should pass platform key")
        if platform not in ('aws', 'linux'):
            raise ValueError("Should use valid platform. Valid platforms are aws or linux")
        return func(*args, **kwargs)

    return wrapper


py_cron_expression/py_cron.py
from py_cron_expression.validations import *
from py_cron_expression.util import remove_zero, PLATFORMS


class CronJobExpression:
    """
    # Example of job definition:
    # .---------------- minute (0 - 59)
    # |  .------------- hour (0 - 23)
    # |  |  .---------- day of month (1 - 31)
    # |  |  |  .------- month (1 - 12) OR jan,feb,mar,apr ...
    # |  |  |  |  .---- day of week (0 - 6) (Sunday=0 or 7)
    # |  |  |  |  |
    # *  *  *  *  *   command to be executed
    m-59/n  *  *  *  *   command
    """

    def __init__(self):
        self._seconds = None
        self._minutes = None
        self._hours = None
        self._day = None
        self._weekly = None
        self._month = None
        self._year = None

    @date_valid
    def cron_expression(self, **kwargs) -> str:
        """
        :param kwargs: dict
        :return:
        10 20 12 30 4 ? 2020
        S  M  H  D  M W Y
        """
        time = kwargs['time']
        cancel = kwargs['cancel']
        date, time = time.split("T")
        self._year, self._month, self._day = date.split("-")
        self._hours, self._minutes, self._seconds = time.split(":")
        if cancel:
            return "{} {} {} {} ? {}".format(
                remove_zero(self._minutes),
                remove_zero(self._hours),
                remove_zero(self._day),
                remove_zero(self._month),
                self._year
            )

        return "{} {} {} {} {} ? {}".format(
            remove_zero(self._seconds),
            remove_zero(self._minutes),
            remove_zero(self._hours),
            remove_zero(self._day),
            remove_zero(self._month),
            self._year
        )

    @platform_valid
    @minute_valid
    def minute(self, platform: str, minutes: int = None) -> str:
        """
        :param platform:
        :param minutes:
        :return:
        """

        if not self._minutes and not minutes:
            raise ValueError("Should pass datetime object or minutes value")

        expression = PLATFORMS.get(platform)
        expression_list = expression.split(" ")
        expression_list[0] = f"*/{minutes or remove_zero(self._minutes)}"
        return " ".join(expression_list)

    @platform_valid
    @hour_valid
    def hourly(self, platform: str, hours: int = None) -> str:
        """
        :param platform:
        :param hours:
        :return:
        """

        if not self._hours and not hours:
            raise ValueError("Should pass datetime object or hours value")

        expression = PLATFORMS.get(platform)
        expression_list = expression.split(" ")
        expression_list[1] = f"*/{hours or remove_zero(self._hours)}"
        return " ".join(expression_list)

    @platform_valid
    @hour_valid
    def daily(self, platform: str, hours: int = None) -> str:
        """
        :param platform:
        :param hours:
        :return:
        """
        if not self._hours and not hours:
            raise ValueError("Should pass datetime object or hours value")

        expression = PLATFORMS.get(platform)
        expression_list = expression.split(" ")
        expression_list[0] = "0"
        expression_list[1] = f"{hours or remove_zero(self._hours)}"
        return " ".join(expression_list)

    @platform_valid
    @hour_valid
    @minute_valid
    @weekly_valid
    def weekly(self, platform: str, hours: int, minutes: int, weekly: str = None) -> str:
        """
        :param platform: aws or linux
        :param hours: [1...23]
        :param minutes: [1...59]
        :param weekly: Sunday to Saturday
        """
        if not self._weekly and not weekly:
            raise ValueError("Should pass datetime object or weekly value")

        expression_list = PLATFORMS.get(platform).split(" ")
        expression_list[0] = f"{minutes}"
        expression_list[1] = f"{hours}"
        if platform.lower() == "aws":
            expression_list[-2] = weekly
        else:
            expression_list[-1] = weekly

        return " ".join(expression_list)


if __name__ == '__main__':
    cron = CronJobExpression()
    h = cron.weekly(**{"platform": "linux", "hours": 10, "minutes": 1, "weekly": "sunday"})
    print(h)
